view/src/extras/address.js

```diff
@@ -0,0 +1,159 @@
+import React from "react";
+import TextField from "@material-ui/core/TextField";
+import Autocomplete from "@material-ui/lab/Autocomplete";
+import LocationOnIcon from "@material-ui/icons/LocationOn";
+import Grid from "@material-ui/core/Grid";
+import Typography from "@material-ui/core/Typography";
+import { makeStyles } from "@material-ui/core/styles";
+import parse from "autosuggest-highlight/parse";
+import throttle from "lodash/throttle";
+
+function loadScript(src, position, id) {
+  if (!position) {
+    return;
+  }
+
+  const script = document.createElement("script");
+  script.setAttribute("async", "");
+  script.setAttribute("id", id);
+  script.src = src;
+  position.appendChild(script);
+}
+
+const autocompleteService = { current: null };
+
+const useStyles = makeStyles((theme) => ({
+  icon: {
+    color: theme.palette.text.secondary,
+    marginRight: theme.spacing(2),
+  },
+}));
+
+export default function GoogleMaps(props) {
+  const classes = useStyles();
+  const [value, setValue] = React.useState(props.location);
+  const [inputValue, setInputValue] = React.useState("");
+  const [options, setOptions] = React.useState([]);
+  const loaded = React.useRef(false);
+
+  if (typeof window !== "undefined" && !loaded.current) {
+    if (!document.querySelector("#google-maps")) {
+      loadScript(
+        "https://maps.googleapis.com/maps/api/js?key=AIzaSyDSPAadPOPcdIXimF1DTzKcHcp7TKa0ht8&libraries=places",
+        document.querySelector("head"),
+        "google-maps"
+      );
+    }
+
+    loaded.current = true;
+  }
+
+  const fetch = React.useMemo(
+    () =>
+      throttle((request, callback) => {
+        autocompleteService.current.getPlacePredictions(request, callback);
+      }, 200),
+    []
+  );
+
+  React.useEffect(() => {
+    let active = true;
+
+    if (!autocompleteService.current && window.google) {
+      autocompleteService.current = new window.google.maps.places.AutocompleteService();
+    }
+    if (!autocompleteService.current) {
+      return undefined;
+    }
+
+    if (inputValue === "") {
+      setOptions(value ? [value] : []);
+      return undefined;
+    }
+
+    fetch({ input: inputValue }, (results) => {
+      if (active) {
+        let newOptions = [];
+
+        if (value) {
+          newOptions = [value];
+        }
+
+        if (results) {
+          newOptions = [...newOptions, ...results];
+        }
+
+        setOptions(newOptions);
+      }
+    });
+
+    return () => {
+      active = false;
+    };
+  }, [value, inputValue, fetch]);
+
+  return (
+    <Autocomplete
```

Comment:
So address.js defines a function called GoogleMap which renders an Autocomplete element...  I am honestly pretty confused about what this actually is!  I think it's a function that renders an autocomplete text field for addresses, right?  That's definitely not a GoogleMap, and it's not an address either.  Maybe this file should be called address_autocomplete_field.js, and the function named AddressAutocompleteField?

I'm also not sure what the philosophical difference is between a Component and a function that renders some JSX.  When would you tend to write one vs. the other in React?  Should this just be a Component that encapsulates all of this JSX into its DOM?

Reply:
You're exactly right about the purpose. I renamed the file and the functional component to what you suggested. 

In terms of functional components vs. traditional components. For functional components you have access to hooks which are really nice ways to change states, and functional components gives you access to the custom hooks such as useEffect() and useMemo(). In general when making individual components, I found functional components to be more intuitive. Thoughts?

Reply:
Yeah, they seem pretty good to me too.  I liked your explanation elsewhere that you're using traditional components for the top-level views, and functional components for the, well, components of those views.  Makes sense to me.